### PR TITLE
Part 4: Array slicing.

### PR DESCRIPTION
**Pull request by schweitzpgi**
Implement the slicing operation on an array. The slice can be over elements or
subcomponents of the contiguous base array. The result is the logical projection
of the original array as a new array. This projection may be stored in a temporary
array.
Implements vector-subscripts in array sectioning/slicing.
Bonus: fir.zero_bits Op.

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -1979,23 +2160,84 @@ class ArrayExprLowering {
...
+                           mlir::Type &recTy,
+                           const Fortran::evaluate::DataRef &dr) {
+    std::visit(Fortran::common::visitors{
+                   [&](const Fortran::evaluate::Component &c) {
+                     buildComponentsPath(components, recTy, c.base());
+                     auto loc = getLoc();
+                     auto name = toStringRef(c.GetLastSymbol().name());
+                     components.push_back(
+                         builder.create<fir::FieldIndexOp>(loc, name, recTy));
+                     recTy = recTy.cast<fir::RecordType>().getType(name);
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2186]: The propagation/mutation of the type is a bit tricky to follow, why not computing the type for the fieldIndexOp on the spot with `converter.genType(c.base().GetLastSymbol())` ? (although, maybe you would need to remove a fir.SequenceType wrapping coming from that if the base is an ArrayRef).

**Review by jeanPerier — COMMENTED** [review]
Looks pretty nice, especially the handling of vector subscript. After doing that, do you think we will avoid placing the slice in the extented values like you did so far or do you think there is still a use case for having  sliceOp in ExtendedValues ?

File: flang/include/flang/Optimizer/Dialect/FIROps.td [hunk 2]
```diff
@@ -3535,86 +3575,98 @@ def fir_DTEntryOp : fir_Op<"dt_entry", []> {
...
-    std::size_t lenParamOffset();
-    mlir::Operation::operand_range lenParamOperands();
-    mlir::Operation::operand_range shapeOperands();
-    mlir::Operation::operand_range shiftOperands();
-    mlir::Operation::operand_range sliceOperands();
-    unsigned getRank();
+    mlir::Operation::operand_range lenParamOperands() { return lenParams(); }
+    mlir::Operation::operand_range shapeOperands() { return shape(); }
+    mlir::Operation::operand_range shiftOperands() { return shift(); }
+    mlir::Operation::operand_range sliceOperands() { return slice(); }
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/include/flang/Optimizer/Dialect/FIROps.td:3616]: All these `fooOperands` methods should be removed eventually.

File: flang/include/flang/Optimizer/Dialect/FIROps.td [hunk 3]
```diff
@@ -3535,86 +3575,98 @@ def fir_DTEntryOp : fir_Op<"dt_entry", []> {
...
     Prior to lowering to LLVM IR dialect, a non-scalar non-trivial embox op will
     be converted to an extended embox. This op will have the following sets of
     arguments.
 
+       - memref: The memory reference being emboxed.
        - shape: A vector that is the runtime shape of the underlying array.
        - shift: A vector that is the runtime origin of the first element.
          The default is a vector of the value 1.
        - slice: A vector of triples that describe an array slice.
+       - subcomponent: A vector of indices for subobject slicing.
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/include/flang/Optimizer/Dialect/FIROps.td:3591]: Field type values are fine. End-to-end support for them is lacking so far, of course.

File: flang/include/flang/Optimizer/Dialect/FIROps.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/include/flang/Optimizer/Dialect/FIROps.td:3622]: OK, sounds good to me.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:2186]: It feels more natural to me to determine the type of a field using what has already been reified in the FIR code than to go back to the front-end and re-derive the type from another representation.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 4]
```diff
@@ -1955,11 +1965,144 @@ class ArrayExprLowering {
...
+                          trips.push_back(one);
+                        },
+                        [&](auto) { useInexactRange(); });
+                  } else {
+                    useInexactRange();
+                  }
+                } else {
+                  // A regular scalar index, which does not yield an array
+                  // section. Use a degenerate slice operation `(e:e:1)` in this
+                  // dimension as a placeholder.
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:2090]: Right. This was sort of a hacky attempt to create a dummy marker in the slice op. I agree that it looks like another solution would be nicer here.

File: flang/lib/Lower/FIRBuilder.cpp [hunk 5]
```diff
@@ -232,10 +233,50 @@ Fortran::lower::FirOpBuilder::createShape(mlir::Location loc,
...
-      });
+      [&](auto) -> mlir::Value { fir::emitFatalError(loc, "not an array"); });
+}
+
+mlir::Value Fortran::lower::FirOpBuilder::createSlice(
+    mlir::Location loc, const fir::ExtendedValue &exv, mlir::ValueRange triples,
+    mlir::ValueRange path) {
+  if (triples.empty()) {
+    // If there is no slicing by triple notation, then take the whole array.
+    llvm::SmallVector<mlir::Value, 8> trips;
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/FIRBuilder.cpp:244]: This looks like an oversight/typo to me as well.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 6]
```diff
@@ -1955,11 +1965,144 @@ class ArrayExprLowering {
...
+  ignoreEvConvert(const A &x) {
+    return Fortran::evaluate::AsGenericExpr(Fortran::common::Clone(x));
+  }
+
+  // Get the `Se::Symbol*` for the subscript expression, `x`. This symbol can be
+  // used to determine the lbound, ubound of the vector.
+  template <typename A>
+  static const Fortran::semantics::Symbol *
+  extractSubscriptSymbol(const Fortran::evaluate::Expr<A> &x) {
+    return std::visit([&](const auto &v) { return extractSubscriptSymbol(v); },
```
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:1995]: OK, I missed that. If you still have some concrete example in mind of this filtering that makes it differ from a raw `UnwrapWholeSymbolDataRef`, that might be useful to add it in comment.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:2186]: OK, fine with me too.

File: flang/test/Lower/arrexp-slice.f90 [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  ! CHECK: %[[shape10:.*]] = fir.shape %c10
+  ! CHECK: %[[mem:.*]] = fir.allocmem !fir.array<3xi32>
+  ! CHECK: %[[shape:.*]] = fir.shape %c3
+  ! CHECK: %[[slice:.*]] = fir.slice %
+  ! CHECK: fir.array_coor %[[A]](%[[shape10]]) [%[[slice]]] %
+  ! CHECK: fir.array_coor %[[mem]](%[[shape]]) %
+  print*, A(1:3:1)
+  call s
+  call i
+end program main
```
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/test/Lower/arrexp-slice.f90:55]: Great, thanks for the test case.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:2090]: Do you plan to solve this now, or do you want to make it a TODO/FIXME to keep track of that ?

**Review by jeanPerier — APPROVED** [review]

**Comment by schweitzpgi** [comment]
There is still work to do with respect to slicing, but I'll tackle more of it in subsequent PRs.  Thanks for the review.

**Pull request merged**